docs/build/use-cases/key-generation.mdx

```diff
@@ -0,0 +1,626 @@
+  wallet: Wallet,
+  to: string,
+  amountEth: string
+): Promise<TransactionReceipt> {
+  const tx = await wallet.sendTransaction({
+    to,
+    value: parseEther(amountEth)
+  });
+  const receipt = await tx.wait();
+  if (receipt == null) {
+    throw new Error("Transaction dropped or replaced before confirmation");
+  }
+  return receipt;
+}
+
+export async function deployContract(
+  wallet: Wallet,
+  abi: any[],
+  bytecode: string,
+  args: unknown[] = []
+): Promise<{ address: string; receipt: TransactionReceipt }> {
+  const factory = new ContractFactory(abi, bytecode, wallet);
+  const contract = await factory.deploy(...args);
+  const deployTx = contract.deploymentTransaction();
+  const receipt = await deployTx?.wait();
+  await contract.waitForDeployment();
+  if (!receipt) {
+    throw new Error("Deployment TX not mined");
+  }
+  return { address: contract.target as string, receipt };
+}
+```
+
+</details>
+
+### `src/keys.ts` — tiny helpers
+
+<details>
+  <summary>src/keys.ts</summary>
+
+```ts
+import { Wallet, getAddress } from "ethers";
+
+export function secretKeyToWallet(skHex: string): Wallet {
+  return new Wallet(skHex);
+}
+
+export function checksumAddress(addr: string): string {
+  return getAddress(addr);
+}
+```
+
+</details>
+
+### `src/scripts/smoke-test.ts` — single end‑to‑end flow
+
+This script prints the App ID (inside ROFL), address, a signed message,
+waits for funding, and deploys the counter contract.
+
+<details>
+  <summary>src/scripts/smoke-test.ts</summary>
+
+```ts
+import "dotenv/config";
+import { readFileSync } from "node:fs";
+import { join } from "node:path";
+import { getAppId, getEvmSecretKey } from "../appd.js";
+import { secretKeyToWallet, checksumAddress } from "../keys.js";
+import { makeProvider, signPersonalMessage, sendEth, deployContract } from "../evm.js";
+import { formatEther, JsonRpcProvider } from "ethers";
+
+const RPC_URL = process.env.BASE_RPC_URL ?? "https://sepolia.base.org";
+const CHAIN_ID = Number(process.env.BASE_CHAIN_ID ?? "84532");
+const KEY_ID = process.env.KEY_ID ?? "evm:base:sepolia";
+
+function sleep(ms: number): Promise<void> {
+  return new Promise((r) => setTimeout(r, ms));
+}
+
+async function waitForFunding(
+  provider: JsonRpcProvider,
+  addr: string,
+  minWei: bigint = 1n,
+  timeoutMs = 15 * 60 * 1000,
+  pollMs = 5_000
+): Promise<bigint> {
+  const start = Date.now();
+  while (Date.now() - start < timeoutMs) {
+    const bal = await provider.getBalance(addr);
+    if (bal >= minWei) return bal;
+    console.log(`Waiting for funding... current balance=${formatEther(bal)} ETH`);
+    await sleep(pollMs);
+  }
+  throw new Error("Timed out waiting for funding.");
+}
+
+async function main() {
+  const appId = await getAppId().catch(() => null);
+  console.log(`ROFL App ID: ${appId ?? "(unavailable outside ROFL)"}`);
+
+  const sk = await getEvmSecretKey(KEY_ID);
+  const wallet = secretKeyToWallet(sk).connect(makeProvider(RPC_URL, CHAIN_ID));
```

Comment:
I think the example should mention that we trust the provider here, and ideally (in a production example) a light-client (e.g., Helios) would be used when fetching state of the remote blockchain.